### PR TITLE
Build script refactor

### PR DESCRIPTION
**Pull request by cyqsimon**
The original intent behind this PR is to accommodate for #2747 (PR: #2755), which requires significant code additional to the build script. Hence it would make sense to clean up the build script beforehand and make it organizationally amenable to said addition (i.e. splitting the code into multiple files and modules). There should be no functional changes.

While doing so, I also did some maintenance work:
- cleaned out the old compgen code that's unlikely to ever be used again
- use `anyhow` to make life a bit easier
- fix inaccurate use of the `option_env!` and `env!` macros (the environment variables should be read at build script run time, not build script compile time)
- use newer code style, specifically
  - implicit variable capture in format string ([since 1.58.0](https://blog.rust-lang.org/2022/01/13/Rust-1.58.0.html#captured-identifiers-in-format-strings))
  - use `[].into_iter().collect()` to immutably construct a `HashMap` ([since 1.56.0 & 2021 edition](https://doc.rust-lang.org/edition-guide/rust-2021/IntoIterator-for-arrays.html)) 
- only rerun build script when relevant source files/env change

### CODE REVIEW

**Review by Enselic — APPROVED** [review]
LGTM! I will merge this after the weekend unless other maintainers raise concerns.

**Review by keith-hall — APPROVED** [review]

**Comment by Enselic** [comment]
With two maintainer approvals, let's merge already now. I don't want to hold you back unnecessarily.

**Pull request merged**
